wallet-next/src/bin/pwalletd.rs

```diff
@@ -1,22 +1,210 @@
+            // Retrieve asset list
+            assets(&client, &mut storage).await?;
+
+            // TODO: this is just a sqlite usage stub
+
+            let row = storage.insert_table().await?;
+            let x = storage.read_table().await?;
+
+            println!(
+                "Hello, pwalletd! I got stuff from sqlite: row {} value {}",
+                row, x
+            );
+
+            // TODO: start gRPC service and respond to command requests
+            // let wallet_server = tokio::spawn(
+            //     Server::builder()
+            //         .trace_fn(|req| match remote_addr(req) {
+            //             Some(remote_addr) => tracing::error_span!("wallet_query", ?remote_addr),
+            //             None => tracing::error_span!("wallet_query"),
+            //         })
+            //         .add_service(WalletServer::new(storage.clone()))
+            //         .serve(
+            //             format!("{}:{}", host, wallet_query_port)
+            //                 .parse()
+            //                 .expect("this is a valid address"),
+            //         ),
+            // );
+        }
+    }
+
+    Ok(())
+}
+
+pub async fn sync(client: &ObliviousQueryClient<Channel>, storage: &mut Storage) -> Result<()> {
+    tracing::info!("starting client sync");
+
+    let start_height = storage
+        .last_block_height()
+        .await
+        .map(|h| h + 1)
+        .unwrap_or(0);
+    let mut stream = client
+        .compact_block_range(tonic::Request::new(CompactBlockRangeRequest {
+            start_height,
+            end_height: 0,
+            chain_id: storage
+                .chain_id()
+                .await
+                .ok_or_else(|| anyhow::anyhow!("missing chain_id"))?,
+        }))
+        .await?
+        .into_inner();
+
+    let mut count = 0;
+    while let Some(block) = stream.message().await? {
+        storage.scan_block(block.try_into()?)?;
+        // very basic form of intermediate checkpointing
+        count += 1;
+        if count % 1000 == 1 {
+            storage.commit()?;
+            tracing::info!(height = ?storage.last_block_height().await.unwrap(), "syncing...");
+        }
+    }
+
+    storage.prune_timeouts();
+    storage.commit()?;
+    tracing::info!(end_height = ?storage.last_block_height().await.unwrap(), "finished sync");
+    Ok(())
+}
+
+pub async fn assets(client: &ObliviousQueryClient<Channel>, storage: &mut Storage) -> Result<()> {
+    // Update asset registry.
+    let request = tonic::Request::new(AssetListRequest {
+        chain_id: storage.chain_id().await.unwrap_or_default(),
+    });
+    let assets: KnownAssets = client.asset_list(request).await?.into_inner().try_into()?;
+    for asset in assets.0 {
+        storage
+            .asset_cache_mut()
+            .extend(std::iter::once(asset.denom));
+    }
+
+    storage.commit()?;
+    tracing::info!("updated asset registry");
+    Ok(())
+}
```

Comment:
This method can be deleted, since we don't currently record any information about asset IDs and denominations in the state we have (it's all public state, anyways).